liquibase-integration-tests/src/test/java/liquibase/dbtest/AbstractIntegrationTest.java

```diff
@@ -1379,6 +1427,39 @@ protected void runUpdate(String changelog) throws CommandExecutionException {
         commandScope.execute();
     }
 
+    /**
+     * An {@link JdbcExecutor} that provides its own {@link DatabaseConnection} and commits after executing
+     * each change
+     */
+    public static class AlternateConnectionExecutor extends JdbcExecutor {
+
+        public AlternateConnectionExecutor() throws Exception {
+            DatabaseTestSystem testSystem = (DatabaseTestSystem) Scope.getCurrentScope().getSingleton(TestSystemFactory.class).getTestSystem("h2alt");
+            testSystem.start();
+            DatabaseConnection connection = new JdbcConnection(testSystem.getConnection());
+            database = DatabaseFactory.getInstance().findCorrectDatabaseImplementation(connection);
```

Comment:
I think instead of defining a new test system and updating an existing one, maybe  we can just do something like this:

```
Database database =
                DatabaseFactory.instance.openDatabase(h2.getConnectionUrl().replace("lbcat", "lbcat2"), h2.getUsername(), h2.getPassword(), null, null)
```
Then we don't need to create H2AltTestSystem, and update H2TestSystem, liquibase.extension.testing..., and liquibase.sdk.yaml.

Another thing I was thinking is instead of having a test in the **AbstractIntegrationTest** suite, we can either move it to **H2IntegrationTest** suite or what I thought could be a good idea is to have a separate `executor` directory where we place a new test class (`CustomExecutorTest`). 

Something like this:
```
@LiquibaseIntegrationTest
class CustomExecutorTest extends Specification {

    @Shared
    public DatabaseTestSystem h2 = Scope.currentScope.getSingleton(TestSystemFactory).getTestSystem("h2") as DatabaseTestSystem

    /**
     * Verifies that the {@link Executor#execute(liquibase.change.Change)} method is called by testing an Executor implementation
     * that uses a separate database connection and commits after each change completes.
     *
     * @throws Exception
     */
    def "validate custom executor is invoked per change"() throws Exception {
        given:
        assumeNotNull(h2.getDatabaseFromFactory())
        String tableName = "test_numbers"
        try {
            CommandUtil.runUpdate(h2, "changelogs/common/runWith.executor.changelog.xml")
        } catch (Exception e) {
            // ok - expect a failure to prove that the changes earlier in the change set committed
        }

        when:
        def alternateConnectionExecutor = (AlternateConnectionExecutor) Scope.getCurrentScope().getSingleton(ExecutorService.class).getExecutor("h2alt", h2.getDatabaseFromFactory())
        Scope.getCurrentScope().getSingleton(ExecutorService.class).setExecutor("jdbc", h2.getDatabaseFromFactory(), alternateConnectionExecutor)

        // Rollback anything in progress on the connection to ensure the executor actually committed
        alternateConnectionExecutor.getDatabase().rollback()

        // Confirm the number of expected rows were inserted into the table in the alt schema
        RowCountPrecondition precondition = new RowCountPrecondition()
        precondition.setSchemaName(h2.getAltSchema())
        precondition.setTableName(tableName)
        precondition.setExpectedRows(2L)
        try {
            precondition.check(alternateConnectionExecutor.getDatabase(), null, null, null)
        } catch (PreconditionFailedException e) {
            fail(e.getFailedPreconditions().get(0).getMessage())
        }

        // Confirm the table was not created in default database and schema
        TableExistsPrecondition tableExistsPrecondition = new TableExistsPrecondition()
        tableExistsPrecondition.setTableName(tableName)
        PreconditionFailedException ex = assertThrows(PreconditionFailedException.class, { ->
            tableExistsPrecondition.check(h2.getDatabaseFromFactory(), null, null, null) })

        then:
        ex.getFailedPreconditions().size() == 1
        String expectedMessage = String.format("%s does not exist", tableName)
        ex.getFailedPreconditions().get(0).getMessage().endsWith(expectedMessage)
    }
```
and in there the inner `AlternateConnectionExecutor` class you created. I tried this, but I was getting some error. Could you try something like that? 

Other than that test looks good to me.

Thanks,
Daniel.

Reply:
I tried the spock approach but I'm afraid I haven't worked much with it and ran into issues. I ended up going with the H2IntegrationTest approach you mentioned. 
Thanks,
David

Reply:
No worries, I think having coverage for both scenarios is good enough. Thank you for addressing these changes so quickly. 

Thanks,
Daniel.